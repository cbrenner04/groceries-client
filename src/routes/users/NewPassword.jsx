import React, { useState } from 'react';
import { Link, useNavigate } from 'react-router-dom';
import { Form } from 'react-bootstrap';
import { toast } from 'react-toastify';

import { EmailField } from '../../components/FormFields';
import axios from '../../utils/api';
import FormSubmission from '../../components/FormSubmission';

export default function NewPassword() {
  const [email, setEmail] = useState('');
  const navigate = useNavigate();

  const handleSubmit = async (event) => {
    event.preventDefault();
    try {
      await axios.post(`/auth/password`, {
        email,
        redirect_url: `${process.env.REACT_APP_PUBLIC_URL}/users/password/edit`,
      });
    } catch {
      // noop
    } finally {
      toast(`If ${email} is in our system, you will receive an email shortly with reset instructions.`, {
        type: 'info',
      });
      navigate('/users/sign_in');
    }
  };

  return (
    <>
      <h2>Forgot your password?</h2>
      <Form onSubmit={handleSubmit} autoComplete="off">
        <EmailField value={email} handleChange={({ target: { value } }) => setEmail(value)} />
        <FormSubmission submitText="Send me reset password instructions" displayCancelButton={false} />
      </Form>
      <Link to="/users/sign_in">Log in</Link>
    </>
  );
}
