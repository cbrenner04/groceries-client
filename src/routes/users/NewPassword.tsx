import React, { type ChangeEvent, type FormEvent, useState } from 'react';
import { Link, useNavigate } from 'react-router';
import { Form } from 'react-bootstrap';
import { toast } from 'react-toastify';

import { EmailField } from 'components/FormFields';
import axios from 'utils/api';
import FormSubmission from 'components/FormSubmission';

const NewPassword: React.FC = (): React.JSX.Element => {
  const [email, setEmail] = useState('');
  const navigate = useNavigate();

  const handleSubmit = async (event: FormEvent<HTMLFormElement>): Promise<void> => {
    event.preventDefault();
    try {
      await axios.post('/auth/password', {
        email,
        redirect_url: `${process.env.REACT_APP_PUBLIC_URL}/users/password/edit`,
      });
    } catch {
      // noop
    } finally {
      toast(`If ${email} is in our system, you will receive an email shortly with reset instructions.`, {
        type: 'info',
      });
      navigate('/users/sign_in');
    }
  };

  return (
    <React.Fragment>
      <h2>Forgot your password?</h2>
      <Form onSubmit={handleSubmit} autoComplete="off">
        <EmailField
          value={email}
          handleChange={(event: ChangeEvent<HTMLInputElement>): void => setEmail(event.target.value)}
        />
        <FormSubmission submitText="Send me reset password instructions" />
      </Form>
      <Link to="/users/sign_in">Log in</Link>
    </React.Fragment>
  );
};

export default NewPassword;
