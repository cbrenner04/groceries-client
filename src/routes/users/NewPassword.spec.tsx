import React from 'react';
import { render, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { toast } from 'react-toastify';
import userEvent from '@testing-library/user-event';

import NewPassword from './NewPassword';
import axios from '../../utils/api';

jest.mock('react-toastify', () => ({
  toast: jest.fn(),
}));

const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

function setup() {
  const user = userEvent.setup();
  const component = render(
    <MemoryRouter>
      <NewPassword />
    </MemoryRouter>,
  );

  return { ...component, user };
}

describe('NewPassword', () => {
  it('requests new password', async () => {
    axios.post = jest.fn().mockResolvedValue({});
    const { container, findByLabelText, findByRole, user } = setup();

    expect(container).toMatchSnapshot();

    await user.type(await findByLabelText('Email'), 'foo@example.com');
    await user.click(await findByRole('button'));
    await waitFor(() => expect(axios.post).toHaveBeenCalledTimes(1));

    expect(toast).toHaveBeenCalledWith(
      'If foo@example.com is in our system, you will receive an email shortly with reset instructions.',
      { type: 'info' },
    );
    expect(mockNavigate).toHaveBeenCalledWith('/users/sign_in');
  });

  it('has same outcome when error occurs', async () => {
    axios.post = jest.fn().mockRejectedValue({ response: { status: 500 } });
    const { findByLabelText, findByRole, user } = setup();

    await user.type(await findByLabelText('Email'), 'foo@example.com');
    await user.click(await findByRole('button'));
    await waitFor(() => expect(axios.post).toHaveBeenCalledTimes(1));

    expect(toast).toHaveBeenCalledWith(
      'If foo@example.com is in our system, you will receive an email shortly with reset instructions.',
      { type: 'info' },
    );
    expect(mockNavigate).toHaveBeenCalledWith('/users/sign_in');
  });
});
