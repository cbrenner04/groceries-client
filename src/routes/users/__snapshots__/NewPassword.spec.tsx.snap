// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewPassword requests new password 1`] = `
<div>
  <h2>
    Forgot your password?
  </h2>
  <form
    autocomplete="off"
    class=""
  >
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="email"
      >
        Email
      </label>
      <input
        class="form-control"
        id="email"
        name="email"
        placeholder="jane.smith@example.com"
        type="email"
        value=""
      />
    </div>
    <div
      class="d-grid gap-2 mt-3"
    >
      <button
        class="btn btn-success"
        type="submit"
      >
        Send me reset password instructions
      </button>
    </div>
  </form>
  <a
    data-discover="true"
    href="/users/sign_in"
  >
    Log in
  </a>
</div>
`;
